### PR TITLE
Support Kafka 0.10 and bump version

### PR DESCRIPTION
**Pull request by dangoldin**
Fix https://github.com/pinterest/secor/issues/258

Just bumping the version ended up breaking a test due to a change in the method signature so I fixed that up and confirmed the tests still pass.

This ended up improving the performance when consuming from Kafka 0.10 brokers significantly.

There are a bunch of deprecation warnings. I wanted to open this one up first so you can see the changes before I end up going through those.

```
Information:Using javac 1.7.0_51 to compile java sources
Information:Module "secor" was fully rebuilt due to project configuration/dependencies changes
Information:10/8/16, 5:55 PM - Compilation completed successfully with 42 warnings in 6s 36ms
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/tools/TestLogMessageProducer.java
Warning:(19, 30) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(20, 22) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
Warning:(21, 22) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(67, 9) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(67, 37) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(68, 9) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(68, 49) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(95, 13) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
Warning:(95, 53) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/tools/RandomPartitioner.java
Warning:(19, 22) java: kafka.producer.Partitioner in kafka.producer has been deprecated
Warning:(27, 43) java: kafka.producer.Partitioner in kafka.producer has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/uploader/Uploader.java
Warning:(133, 35) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(158, 58) java: getKafkaPartition() in com.pinterest.secor.common.LogFilePath has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/io/impl/DelimitedTextFileReaderWriterFactory.java
Warning:(73, 32) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/tools/LogFilePrinter.java
Warning:(44, 38) java: Reader(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.fs.Path,org.apache.hadoop.conf.Configuration) in org.apache.hadoop.io.SequenceFile.Reader has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/io/impl/MessagePackSequenceFileReaderWriterFactory.java
Warning:(68, 28) java: Reader(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.fs.Path,org.apache.hadoop.conf.Configuration) in org.apache.hadoop.io.SequenceFile.Reader has been deprecated
Warning:(116, 44) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class,org.apache.hadoop.io.SequenceFile.CompressionType,org.apache.hadoop.io.compress.CompressionCodec) in org.apache.hadoop.io.SequenceFile has been deprecated
Warning:(120, 44) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class) in org.apache.hadoop.io.SequenceFile has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/io/impl/ProtobufParquetFileReaderWriterFactory.java
Warning:(55, 33) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/io/impl/SequenceFileReaderWriterFactory.java
Warning:(67, 28) java: Reader(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.fs.Path,org.apache.hadoop.conf.Configuration) in org.apache.hadoop.io.SequenceFile.Reader has been deprecated
Warning:(98, 44) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class,org.apache.hadoop.io.SequenceFile.CompressionType,org.apache.hadoop.io.compress.CompressionCodec) in org.apache.hadoop.io.SequenceFile has been deprecated
Warning:(102, 44) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class) in org.apache.hadoop.io.SequenceFile has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/tools/LogFileVerifier.java
Warning:(63, 32) java: getKafkaPartition() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(70, 42) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(130, 65) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(131, 87) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(133, 41) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
/Users/danielgoldin/code/secor/src/main/java/com/pinterest/secor/uploader/HadoopS3UploadManager.java
Information:java: Some input files use unchecked or unsafe operations.
Information:java: Recompile with -Xlint:unchecked for details.
/Users/danielgoldin/code/secor/src/test/java/com/pinterest/secor/performance/PerformanceTest.java
Warning:(49, 30) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(50, 22) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
Warning:(51, 22) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(100, 9) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(100, 45) java: kafka.producer.ProducerConfig in kafka.producer has been deprecated
Warning:(102, 9) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(102, 49) java: kafka.javaapi.producer.Producer in kafka.javaapi.producer has been deprecated
Warning:(115, 17) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
Warning:(115, 57) java: kafka.producer.KeyedMessage in kafka.producer has been deprecated
/Users/danielgoldin/code/secor/src/test/java/com/pinterest/secor/io/FileReaderWriterFactoryTest.java
Warning:(150, 33) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class) in org.apache.hadoop.io.SequenceFile has been deprecated
Warning:(162, 33) java: createWriter(org.apache.hadoop.fs.FileSystem,org.apache.hadoop.conf.Configuration,org.apache.hadoop.fs.Path,java.lang.Class,java.lang.Class,org.apache.hadoop.io.SequenceFile.CompressionType,org.apache.hadoop.io.compress.CompressionCodec) in org.apache.hadoop.io.SequenceFile has been deprecated
/Users/danielgoldin/code/secor/src/test/java/com/pinterest/secor/common/LogFilePathTest.java
Warning:(67, 50) java: getKafkaPartition() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(68, 56) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(75, 51) java: getKafkaPartition() in com.pinterest.secor.common.LogFilePath has been deprecated
Warning:(76, 57) java: getOffset() in com.pinterest.secor.common.LogFilePath has been deprecated
```

### CODE REVIEW

**Comment by dangoldin** [comment]
This ends up improving the performance from Kafka 0.10 brokers significantly.

**Comment by HenryCaiHaiying** [comment]
Thanks for the contribution.  It looks like the changes needed for kafka 10 is actually quite small.  But I don't want to remove the kafka 8 at this moment since lots of people are still using that version.  I think the way to make both kafka 8 and kafka 10 co-exist in by playing the trick of profile in pom.xml.

E.g.
<profile>
    <id>kafka10</id>
    <dependencies>
        <dependency>
            <groupId>org.apache.kafka</groupId>
             <version>0.10.0.1</version>
And create another profile for kafka8, use '-P' flag for mvn clean package to specify which profile to invoke

**Comment by dangoldin** [comment]
Hey - that makes sense but I'm pretty new to profiles and duplicating the <build> section feels dirty. Is there a way to reuse that one but just have two different <dependencies> sections?

**Comment by dangoldin** [comment]
The other issue is that my updates to PerformanceTest.java break under Kafka 0.8 with the following errors:

- kafka.admin.RackAwareMode does not exist in 0.8
- ZkUtils is private in 0.8 but made public in 0.10

I can see if there's a way to write the test that it works successfully in both.

**Comment by HenryCaiHaiying** [comment]
If you profile in pom.xml, there is no duplicate definition of
dependencies.  You will have a structure like the following (I used the
example of our other project which needs to compile under hadoop 1 or
hadoop2):

<dependencies>
   ... all the common dependencies between kafka 8 and kafka 10 ...
<profiles>
   <profile>kafka8
       <dependencies> kafka 8 specific dependencies, e.g. the kafka 0.8.2.1
   <profile>kafka10
       <dependencies> kafka 10 specific dependencies e.g. kafka 0.10.0.1
       <build>

                <plugins>

                    <plugin>

                        <groupId>org.apache.maven.plugins</groupId>

                        <artifactId>maven-compiler-plugin</artifactId>

                        <configuration>
<testExclude>com/pinterest/hadoop/mapreduce/TestCombineTextInputFormat.java</testExclude>

                            <excludes>

                                <comments>These following source files
depends on hadoop 1.0 APIs. So we need to exclude them here.</comments>

  <exclude>com/pinterest/hadoop/hbase/HBaseCompleteLoad.java</exclude>

**Comment by smferguson** [comment]
@dangoldin curious. any progress here? my company needs 0.10 support as well.

**Comment by dangoldin** [comment]
This branch works and we're running it in production. All that's left is reconciling the unit test to work for both versions.

**Comment by smferguson** [comment]
cool. is reconciling the unit test on your radar?

**Comment by dangoldin** [comment]
Yep - just taking a bit of time since I want to support both versions with the same code.

**Comment by dangoldin** [comment]
Just tweaked the code to add support for Kafka 0.10 with a separate build profile and updated the readme on how to pick it.

**Comment by dangoldin** [comment]
Hmm - seems to be failing due to the gpg plugin but I took it out of the kafka-0.10 profile. Is there something else I should do here?

File: pom.xml [hunk 1]
```diff
@@ -532,6 +521,9 @@
     <profiles>
         <profile>
             <id>release</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
```
Comment (HenryCaiHaiying) [inline pom.xml]: Maybe this is why gpg was brought in.  release profile does not needs to be activated by default.  Does it require you have to a default profile?

**Comment by dangoldin** [comment]
Unfortunately if I remove that activeByDefault then it requires to be built by specifying a profile: "mvn -Prelease install"

I tried to preserve the existing behavior by having the release be the default profile but it seems as if that's causing the gpg package to be pulled in which is failing in the test phase.

**Comment by HenryCaiHaiying** [comment]
How about you add another profile "kafka8-dev" (or just called 'kafka8') and make that profile as the default profile.  That profile will not include gpg and other released related stuff.

**Comment by dangoldin** [comment]
@HenryCaiHaiying - That seemed to have done the trick. Thanks.

There are 3 profiles now with the kafka-0.8-dev and release having the same config settings except that the 0.8 version doesn't pull in gpg. There's also a kakfa-0.10 profile containing the 0.10 consumer code.

File: pom.xml [hunk 2]
```diff
@@ -576,6 +565,114 @@
...
+        </profile>
+        <profile>
+            <id>kafka-0.8-dev</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
+            </activation>
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
```
Comment (HenryCaiHaiying) [inline pom.xml]: Do we need these two maven plugin (attach-source, javadoc) for kafka-0.8-dev profile?  I thought they were only used during release build.

**Comment by dangoldin** [comment]
You're right - I just copied them from the release but they're not necessary. Removed them from both the 0.8 and 0.10 dev profiles.

**Comment by dangoldin** [comment]
Looks as if the CI build isn't kicking off automatically though. Is it just being throttled?

**Comment by HenryCaiHaiying** [comment]
Thanks for your patience.  This is a lot of efforts putting into this PR.

**Comment by dangoldin** [comment]
All good and thanks for the pointers.

**Comment by smferguson** [comment]
@dangoldin thanks! this makes life a bit easier for me

**Pull request merged**